ls1intum/Artemis
# [Bugfix] Show resolve repo conflict in assessment view

### Checklist

### Motivation and Context


If the repository on Artemis is in a conflict state, a tutor cannot resolve the conflict in the assessment view. Whereas as a student, there is a resolve conflict button available, but not as a tutor for assessment.
Fixes: #2990 #2885 

### Description

We introduce a new RepositoryActionType `RESET` which ist used if a repository shall be reset. 
A student is only allowed to reset the repository if the participation is not locked yet and the exam is not over. 
A tutor/instructor is only allowed to reset a students repository if the participation is locked or the exam is over.

### Steps for Testing


1. Log in to Artemis
2. Navigate to Course Administration
3. Create a programming exercise with manual assessment
4. Participate in the exercise and produce a conflict (e.g. first, make changes in the Online Code Editor, and click `Save` NOT `Submit` yet. Make local changes and push code from a local Git client. Then submit the code in the online editor.)
5. You should immediately get a Conflict Warning.
6. Assess this exercise and observe the Conflict Warning and that no files are displayed in the left file browser. There should be a red button to Resolve the Repository. After clicking it, the file browser should list the files again
7. Also test this scenario as an exam programming exercise

EDIT:
Note: Please keep in mind that we need to also test the regular behaviour of the online code editor (as student, tutor, instructor, in course, in exam) to verify everything is still working as expeced. 
 
### Test Coverage



RepositoryProgrammingExerciseParticipationResource.java: 76% (Rest are throws)

### Screenshots


#### Assessment view with repo in conflict state:
![image](https://user-images.githubusercontent.com/33484318/110526266-dd2de180-8115-11eb-8be7-7d527769eb50.png)

#### After resolving the conflict: 
![image](https://user-images.githubusercontent.com/33484318/110539616-c2fbff80-8125-11eb-9a96-1dbed64ec1b4.png)
